### PR TITLE
Rustdoc examples

### PR DESCRIPTION
**Pull request by Hoverbear**
Supersedes https://github.com/pingcap/raft-rs/pull/65.

## Motivation

We're having lots of people who want to use Raft, but they don't know how! @queenypingcap did a great job assembling some examples and writing around how to use the crate, however it made the README very long.

This PR moves that content (with minor refactoring) into the Rust API docs generated by https://docs.rs/raft/ and `cargo doc`.

## Further Work

There should be a follow up in the future that makes all examples fully tested, but for now this is a big enough step forward. :)

### CODE REVIEW

**Comment by ngaut** [comment]
Good job.

**Comment by queenypingcap** [comment]
LGTM

File: src/lib.rs [hunk 1]
```diff
@@ -25,6 +25,181 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+/*!
```
Comment (BusyJay) [inline src/lib.rs]: Should we add a `#[deny(missing_docs)]`?

Comment (Hoverbear) [reply src/lib.rs]: I'd love to do that eventually but right now that would make the build fail until we added quite some documentation. I opened https://github.com/pingcap/raft-rs/issues/72 to track this.

**Comment by hicqu** [comment]
rest LGTM.

File: tests/cases/test_raft.rs [hunk 2]
```diff
@@ -715,9 +715,12 @@ fn test_leader_cycle_pre_vote() {
 fn test_leader_cycle_with_config(pre_vote: bool) {
     let mut network = Network::new_with_config(vec![None, None, None], pre_vote);
     for campaigner_id in 1..4 {
-        network.send(vec![
-            new_message(campaigner_id, campaigner_id, MessageType::MsgHup, 0),
-        ]);
+        network.send(vec![new_message(
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply tests/cases/test_raft.rs]: Raft uses stable `rustfmt`, not nightly.

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**